### ISSUE TITLE
PK auto-generation breaks after DbUpdateException

### ISSUE THREAD

**Issue by Remleo** [dotnet/efcore#8036]
Code to reproduce the bug:

```c#
            var oneGroup = new ProductGroup
            {
                Name = "One"
            };
            dbContext.Add(oneGroup);

            dbContext.SaveChanges(); // Save to DB
            dbContext.Entry(oneGroup).State = EntityState.Detached; // Detach from context

            // New product
            dbContext.Add(new ProductCard
            {
                Name = "Some product"
            });

            dbContext.Add(new ProductGroup
            {
                Id = oneGroup.Id, // For simulating the error
                Name = "NewOne"
            });

            try
            {
                dbContext.SaveChanges(); // Try save
            }
            catch (DbUpdateException e)
            {
                foreach (var entiry in e.Entries)
                {
                    entiry.State = EntityState.Detached; // Detach entity with error
                }

                dbContext.SaveChanges(); // Save again
            }
```

![2017-03-30_17-15-26](https://cloud.githubusercontent.com/assets/10843339/24508567/7ac489ea-156c-11e7-8953-9e9a61d8c3a0.png)

After re-saving we have ID = -2147482647 in DB for ProductCard with name "Some product".

EF Core version: 1.1.1
Microsoft.EntityFrameworkCore: 1.1.1
Npgsql.EntityFrameworkCore.PostgreSQL: 1.1.0
Operating system: Windows 10

**Comment by ajcvickers** [comment]
@Remleo In the code above, the first entity gets a temporary key value assigned. The key value is then assigned to a new entity, but EF does not know that it is a temporary key value and hence uses it as the real key value.

Can you explain what it is that you are actually trying to achieve when this fails? Why assign the new entity the temporary key value generated for the previous entity?

**Comment by Remleo** [comment]
The problem was discovered by accident. There was other reason for error in my case. Any DB-specific error breaks pk value autogeneration for others entity (even for other entity type).

That example shows how the error with saving entity type of `ProductGroup` breaks pk value autogen.

In other words:

```c#  
            // Adding normal entity with no error (for short "Normal Entity")
            dbContext.Add(new ProductCard
            {
                Name = "Some product"
            });
            // Adding other entity with error (for short "Error Entity")
            dbContext.Add(<Entity with error for DB>);

            try
            {
                dbContext.SaveChanges(); // Try save
            }
            catch (DbUpdateException e)
            {
                foreach (var entiry in e.Entries)
                {
                    entiry.State = EntityState.Detached; // Detach entity with error
                }

                // Now only the "Normal Entity" left for saving
                // Re-saving
                dbContext.SaveChanges(); 
            }
```

After re-saving the autogeneration not working for "Normal Entity". If no error was thrown the "Normal Entity" will have normal PK value.

**Comment by ajcvickers** [comment]
This is a dupe of #6720, which is fixed for 2.0. Until then, you can work around the issue by explicitly marking key values as temporary again before calling SaveChanges. Something like:
```C#
catch (DbUpdateException e)
{
    foreach (var entry in e.Entries)
    {
        entry.State = EntityState.Detached;
    }

    foreach (var entry in context.ChangeTracker.Entries())
    {
        entry.Property("Id").IsTemporary = true;
    }

    context.SaveChanges();
}
```
